src/adapter/etl-adapter-csv/src/Flow/ETL/Adapter/CSV/CSVLineReader.php

```diff
@@ -0,0 +1,92 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Flow\ETL\Adapter\CSV;
+
+use Flow\Filesystem\SourceStream;
+
+final readonly class CSVLineReader
+{
+    /**
+     * @param null|int<1, max> $charactersReadInLine
+     */
+    public function __construct(
+        private string $enclosure,
+        private ?int $charactersReadInLine = null,
+        private bool $removeBOM = true,
+    ) {
+    }
+
+    /**
+     * @return \Generator<int, string>
+     */
+    public function readLines(SourceStream $stream) : \Generator
+    {
+        $lineNumber = 0;
+        $buffer = '';
+
+        foreach ($stream->readLines(length: $this->charactersReadInLine) as $rawLine) {
+            $buffer .= $rawLine;
+
+            if (!\str_contains($buffer, $this->enclosure)) {
+                yield $this->removeBOM && $lineNumber === 0 ? $this->removeBOMFromLine(\rtrim($buffer, "\r\n")) : rtrim($buffer, "\r\n");
```

Comment:
Shouldn't the BOM check be only on line zero? There is no point in repeating the check on every line, no?

Reply:
that's how it works now, it checks if `removeBOM` is needed and if it's needed then it checks if the line number is 0 (first line).